docs/user/datastores/query_interceptor.rst

```diff
@@ -0,0 +1,114 @@
+.. _query_interceptors:
+
+Query Interceptors and Guards
+=============================
+
+GeoMesa provides a chance for custom logic to be applied to a query before executing it. Query interceptors must
+be specified through user data in the simple feature type, and may be set before calling ``createSchema``, or
+updated by calling ``updateSchema``. To indicate query interceptors, use the key ``geomesa.query.interceptors``:
+
+.. code-block:: java
+
+    sft.getUserData().put("geomesa.query.interceptors", "com.example.MyQueryInterceptor");
+
+The value must be a comma-separated string consisting of the names of one or more classes implementing
+the trait ``org.locationtech.geomesa.index.planning.QueryInterceptor``:
+
+.. code-block:: scala
+
+    /**
+      * Provides a hook to modify a query before executing it
+      */
+    trait QueryInterceptor extends Closeable {
+
+      /**
+        * Called exactly once after the interceptor is instantiated
+        *
+        * @param ds data store
+        * @param sft simple feature type
+        */
+      def init(ds: DataStore, sft: SimpleFeatureType): Unit
+
+      /**
+        * Modifies the query in place
+        *
+        * @param query query
+        */
+      def rewrite(query: Query): Unit
+
+      /**
+       * Hook to allow interception of a query after extracting the query values
+       *
+       * @param strategy query strategy
+       * @return an exception if the query should be stopped
+       */
+      def guard(strategy: QueryStrategy): Option[IllegalArgumentException] = None
+    }
```

Comment:
i wonder if we should make this a java interface... and if that would break existing impls or not.

Reply:
Hmm... do we have folks implementing the GeoMesa Java APIs?  Also, it should still be possible to do so...